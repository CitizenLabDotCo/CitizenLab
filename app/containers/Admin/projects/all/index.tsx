import React, { PureComponent } from 'react';
import { isNilOrError } from 'utils/helperUtils';

// services
import { IProjectData, reorderProject } from 'services/projects';
import GetProjects, { GetProjectsChildProps } from 'resources/GetProjects';
import GetProjectGroups from 'resources/GetProjectGroups';

// localisation
import { FormattedMessage } from 'utils/cl-intl';
import T from 'components/T';
import messages from '../messages';

// components
import { SortableList, SortableRow, List, Row } from 'components/admin/ResourceList';
import PageWrapper from 'components/admin/PageWrapper';
import Button from 'components/UI/Button';
import { PageTitle, SectionSubtitle } from 'components/admin/Section';
import StatusLabel from 'components/UI/StatusLabel';
import HasPermission from 'components/HasPermission';
import InfoTooltip from 'components/admin/InfoTooltip';

// style
import { fontSizes, colors } from 'utils/styleUtils';
import styled from 'styled-components';

const ListHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  margin-bottom: 25px;

  &.marginTop {
    margin-top: 70px;
  }
`;

const ListHeaderTitle = styled.h3`
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.xl}px;
  font-weight: 400;
  padding: 0;
  margin: 0;
  margin-right: 7px;
`;

const RowContent = styled.div`
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const RowContentInner = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  margin-right: 20px;
`;

const RowTitle = styled(T)`
  font-size: ${fontSizes.base}px;
  font-weight: 400;
  line-height: 24px;
  margin-right: 10px;
`;

const StyledStatusLabel = styled(StatusLabel)`
  margin-right: 5px;
  margin-top: 4px;
  margin-bottom: 4px;
`;

const StyledButton = styled(Button)``;

interface InputProps { }

interface DataProps {
  projects: GetProjectsChildProps;
}

interface Props extends InputProps, DataProps { }

interface State {
  publishedProjects: IProjectData[] | null;
  draftProjects: IProjectData[] | null;
  archivedProjects: IProjectData[] | null;
}

class AdminProjectsList extends PureComponent<Props, State> {

  handleReorder = (projectId, newOrder) => {
    reorderProject(projectId, newOrder);
  }

  render() {
    const { projects } = this.props;
    let lists: JSX.Element | null = null;

    if (projects && !isNilOrError(projects.projectsList)) {
      const { projectsList } = projects;
      const publishedProjects = projectsList.filter((project) => {
        return project.attributes.publication_status === 'published';
      });
      const draftProjects = projectsList.filter((project) => {
        return project.attributes.publication_status === 'draft';
      });
      const archivedProjects = projectsList.filter((project) => {
        return project.attributes.publication_status === 'archived';
      });

      const row = (project: IProjectData) => {
        return (
          <RowContent>
            <RowContentInner className="expand primary">
              <RowTitle value={project.attributes.title_multiloc} />
              {project.attributes.visible_to === 'groups' &&
                <GetProjectGroups projectId={project.id}>
                  {(projectGroups) => {
                    if (!isNilOrError(projectGroups)) {
                      return (
                        <StyledStatusLabel
                          text={projectGroups.length > 0 ? (
                            <FormattedMessage {...messages.xGroupsHaveAccess} values={{ groupCount: projectGroups.length }} />
                          ) : (
                              <FormattedMessage {...messages.onlyAdminsCanView} />
                            )}
                          color="clBlue"
                          icon="lock"
                        />
                      );
                    }

                    return null;
                  }}
                </GetProjectGroups>
              }

              {project.attributes.visible_to === 'admins' &&
                <StyledStatusLabel
                  text={<FormattedMessage {...messages.onlyAdminsCanView} />}
                  color="clBlue"
                  icon="lock"
                />
              }
            </RowContentInner>
            <StyledButton
              className={`e2e-admin-edit-project ${project.attributes.title_multiloc['en-GB'] ? project.attributes.title_multiloc['en-GB'] : ''} ${project.attributes.process_type === 'timeline' ? 'timeline' : 'continuous'}`}
              linkTo={`/admin/projects/${project.id}/edit`}
              style="secondary"
              circularCorners={false}
              icon="edit"
            >
              <FormattedMessage {...messages.editButtonLabel} />
            </StyledButton>
          </RowContent>
        );
      };

      lists = (
        <>
          {publishedProjects && publishedProjects.length > 0 &&
            <>
              <ListHeader className="marginTop">
                <ListHeaderTitle>
                  <FormattedMessage {...messages.published} />
                </ListHeaderTitle>
                <InfoTooltip {...messages.publishedTooltip} />
              </ListHeader>
              <HasPermission item="projects" action="reorder">
                <SortableList
                  items={publishedProjects}
                  onReorder={this.handleReorder}
                  id="e2e-admin-published-projects-list"
                  className="e2e-admin-projects-list"
                >
                  {({ itemsList, handleDragRow, handleDropRow }) => (
                    itemsList.map((project: IProjectData, index: number) => (
                      <SortableRow
                        className="e2e-admin-projects-list-item"
                        key={project.id}
                        id={project.id}
                        index={index}
                        moveRow={handleDragRow}
                        dropRow={handleDropRow}
                        lastItem={(index === publishedProjects.length - 1)}
                      >
                        {row(project)}
                      </SortableRow>
                    ))
                  )}
                </SortableList>
                <HasPermission.No>
                  <List>
                    {publishedProjects.map((project, index) => (
                      <Row key={project.id} lastItem={(index === publishedProjects.length - 1)}>
                        {row(project)}
                      </Row>
                    ))}
                  </List>
                </HasPermission.No>
              </HasPermission>
            </>
          }

          {draftProjects && draftProjects.length > 0 &&
            <>
              <ListHeader className="marginTop">
                <ListHeaderTitle>
                  <FormattedMessage {...messages.draft} />
                </ListHeaderTitle>
                <InfoTooltip {...messages.draftTooltip} />
              </ListHeader>
              <HasPermission item="projects" action="reorder">
                <SortableList
                  items={draftProjects}
                  onReorder={this.handleReorder}
                  className="e2e-admin-projects-list"
                  id="e2e-admin-draft-projects-list"
                >
                  {({ itemsList, handleDragRow, handleDropRow }) => (
                    itemsList.map((project: IProjectData, index: number) => (
                      <SortableRow
                        key={project.id}
                        id={project.id}
                        className="e2e-admin-projects-list-item"
                        index={index}
                        moveRow={handleDragRow}
                        dropRow={handleDropRow}
                        lastItem={(index === draftProjects.length - 1)}
                      >
                        {row(project)}
                      </SortableRow>
                    ))
                  )}
                </SortableList>
                <HasPermission.No>
                  <List>
                    {draftProjects.map((project, index) => (
                      <Row key={project.id} lastItem={(index === draftProjects.length - 1)}>
                        {row(project)}
                      </Row>
                    ))}
                  </List>
                </HasPermission.No>
              </HasPermission>
            </>
          }

          {archivedProjects && archivedProjects.length > 0 &&
            <>
              <ListHeader className="marginTop">
                <ListHeaderTitle>
                  <FormattedMessage {...messages.archived} />
                </ListHeaderTitle>
                <InfoTooltip {...messages.archivedTooltip} />
              </ListHeader>
              <List id="e2e-admin-archived-projects-list">
                {archivedProjects.map((project, index) => (
                  <Row
                    className="e2e-admin-projects-list-item"
                    key={project.id}
                    lastItem={(index === archivedProjects.length - 1)}
                  >
                    {row(project)}
                  </Row>
                ))}
              </List>
            </>
          }
        </>
      );
    }

    return (
      <>
        <PageTitle>
          <FormattedMessage {...messages.overviewPageTitle} />
        </PageTitle>
        <HasPermission item={{ type: 'route', path: '/admin/projects/new' }} action="access">
          <SectionSubtitle>
            <FormattedMessage {...messages.overviewPageSubtitle} />
          </SectionSubtitle>
        </HasPermission>

        <PageWrapper>
          <HasPermission item={{ type: 'route', path: '/admin/projects/new' }} action="access">
            <ListHeader>
              <Button className="e2e-admin-add-project" linkTo="/admin/projects/new" style="cl-blue" circularCorners={false} icon="plus-circle">
                <FormattedMessage {...messages.addNewProject} />
              </Button>
            </ListHeader>
          </HasPermission>
          {lists}
        </PageWrapper>
      </>
    );
  }
}

export default (inputProps: InputProps) => (
  <GetProjects publicationStatuses={['draft', 'published', 'archived']} filterCanModerate={true}>
    {projects => <AdminProjectsList {...inputProps} projects={projects} />}
  </GetProjects>
);
