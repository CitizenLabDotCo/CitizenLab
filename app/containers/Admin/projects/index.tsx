import * as React from 'react';

// components
import HelmetIntl from 'components/HelmetIntl';

// i18n
import messages from './messages';

type Props = {};

type State = {};

export default class ProjectDashboard extends React.PureComponent<Props, State> {
  render() {
    return (
      <>
        <HelmetIntl title={messages.helmetTitle} description={messages.helmetDescription} />
        {this.props.children}
      </>
    );
  }
}
