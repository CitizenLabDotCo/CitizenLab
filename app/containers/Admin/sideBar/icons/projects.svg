<svg xmlns="http://www.w3.org/2000/svg" width="21" height="18" viewBox="0 0 21 18">
    <path fill="#FFF" fill-rule="nonzero" d="M10.5 0a.801.801 0 0 0-.449.138L.915 6.34a.788.788 0 0 0 .449 1.443H2.87v6.33h-.481a.793.793 0 0 0-.797.79c0 .437.356.79.797.79H18.61c.44 0 .796-.353.796-.79a.792.792 0 0 0-.796-.79h-.482v-6.33h1.505c.35 0 .657-.226.762-.557a.785.785 0 0 0-.312-.886L10.95.138A.8.8 0 0 0 10.5 0zm0 1.747l6.56 4.456H3.939l6.56-4.456zM4.461 7.784h1.824v6.33H4.462v-6.33zm3.416 0h1.826v6.33H7.878v-6.33zm3.418 0h1.824v6.33h-1.824v-6.33zm3.417 0h1.824v6.33h-1.824v-6.33zM.796 16.386a.793.793 0 0 0-.796.79c0 .437.356.79.796.79h19.407c.44 0 .796-.353.796-.79a.793.793 0 0 0-.796-.79H.796z" opacity=".6"/>
</svg>
