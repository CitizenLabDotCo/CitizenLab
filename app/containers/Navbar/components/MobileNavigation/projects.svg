<svg width="32" height="34" viewBox="0 0 32 34" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<title>Group</title>
<desc>Created using Figma</desc>
<g id="Canvas" transform="translate(3684 383)">
<g id="Group">
<g id="Shape">
<use xlink:href="#path0_stroke" transform="translate(-3680.61 -366.136)"/>
</g>
<g id="Shape">
<use xlink:href="#path0_stroke" transform="translate(-3672.99 -366.136)"/>
</g>
<g id="Shape">
<use xlink:href="#path0_stroke" transform="translate(-3665.38 -366.136)"/>
</g>
<g id="Shape">
<use xlink:href="#path0_stroke" transform="translate(-3657.76 -366.136)"/>
</g>
<g id="Rectangle-path">
<use xlink:href="#path1_stroke" transform="translate(-3682.8 -353.119)"/>
</g>
<g id="Shape">
<use xlink:href="#path2_stroke" transform="translate(-3682.8 -368.847)"/>
</g>
<g id="Shape">
<use xlink:href="#path3_stroke" transform="translate(-3682.8 -374.353)"/>
</g>
<g id="Shape">
<use xlink:href="#path4_stroke" transform="translate(-3668.33 -381.506)"/>
</g>
<g id="Rectangle-path">
<use xlink:href="#path5_stroke" transform="translate(-3667.83 -381.506)"/>
</g>
<g id="Rectangle-path">
<use xlink:href="#path5_stroke" transform="translate(-3664.57 -380.422)"/>
</g>
</g>
</g>
<defs>
<path id="path0_stroke" d="M 0.75 0C 0.75 -0.414214 0.414214 -0.75 0 -0.75C -0.414214 -0.75 -0.75 -0.414214 -0.75 0L 0.75 0ZM 0 13.0169L -0.75 13.0169C -0.75 13.4312 -0.414214 13.7669 1.66533e-16 13.7669L 0 13.0169ZM 2.7193 13.0169L 2.7193 13.7669C 3.13351 13.7669 3.4693 13.4312 3.4693 13.0169L 2.7193 13.0169ZM 3.4693 0C 3.4693 -0.414214 3.13351 -0.75 2.7193 -0.75C 2.30508 -0.75 1.9693 -0.414214 1.9693 0L 3.4693 0ZM -0.75 0L -0.75 13.0169L 0.75 13.0169L 0.75 0L -0.75 0ZM 0 13.7669L 2.7193 13.7669L 2.7193 12.2669L 0 12.2669L 0 13.7669ZM 3.4693 13.0169L 3.4693 0L 1.9693 0L 1.9693 13.0169L 3.4693 13.0169Z"/>
<path id="path1_stroke" d="M 0 0L 0 -0.75C -0.414214 -0.75 -0.75 -0.414214 -0.75 1.66533e-16L 0 0ZM 29.9319 0L 30.6819 0C 30.6819 -0.414214 30.3461 -0.75 29.9319 -0.75L 29.9319 0ZM 29.9319 2.71078L 29.9319 3.46078C 30.3461 3.46078 30.6819 3.12499 30.6819 2.71078L 29.9319 2.71078ZM 0 2.71078L -0.75 2.71078C -0.75 3.12499 -0.414214 3.46078 1.66533e-16 3.46078L 0 2.71078ZM 0 0.75L 29.9319 0.75L 29.9319 -0.75L 0 -0.75L 0 0.75ZM 29.1819 0L 29.1819 2.71078L 30.6819 2.71078L 30.6819 0L 29.1819 0ZM 29.9319 1.96078L 0 1.96078L 0 3.46078L 29.9319 3.46078L 29.9319 1.96078ZM 0.75 2.71078L 0.75 0L -0.75 0L -0.75 2.71078L 0.75 2.71078Z"/>
<path id="path2_stroke" d="M 30.6819 0C 30.6819 -0.414214 30.3461 -0.75 29.9319 -0.75C 29.5176 -0.75 29.1819 -0.414214 29.1819 0L 30.6819 0ZM 29.9319 2.71078L 29.9319 3.46078C 30.3461 3.46078 30.6819 3.12499 30.6819 2.71078L 29.9319 2.71078ZM 0 2.71078L -0.75 2.71078C -0.75 3.12499 -0.414214 3.46078 1.66533e-16 3.46078L 0 2.71078ZM 0.75 0C 0.75 -0.414214 0.414214 -0.75 0 -0.75C -0.414214 -0.75 -0.75 -0.414214 -0.75 0L 0.75 0ZM 29.1819 0L 29.1819 2.71078L 30.6819 2.71078L 30.6819 0L 29.1819 0ZM 29.9319 1.96078L 0 1.96078L 0 3.46078L 29.9319 3.46078L 29.9319 1.96078ZM 0.75 2.71078L 0.75 0L -0.75 0L -0.75 2.71078L 0.75 2.71078Z"/>
<path id="path3_stroke" d="M -0.258965 4.8023C -0.647703 4.94532 -0.846895 5.3764 -0.703873 5.76513C -0.560851 6.15387 -0.129774 6.35306 0.258965 6.21004L -0.258965 4.8023ZM 14.9659 0L 15.2249 -0.703873C 15.0577 -0.765376 14.8741 -0.765376 14.707 -0.703873L 14.9659 0ZM 29.6729 6.21004C 30.0616 6.35306 30.4927 6.15387 30.6357 5.76513C 30.7788 5.3764 30.5796 4.94532 30.1908 4.8023L 29.6729 6.21004ZM 0.258965 6.21004L 15.2249 0.703873L 14.707 -0.703873L -0.258965 4.8023L 0.258965 6.21004ZM 14.707 0.703873L 29.6729 6.21004L 30.1908 4.8023L 15.2249 -0.703873L 14.707 0.703873Z"/>
<path id="path4_stroke" d="M -0.75 7.15336C -0.75 7.56757 -0.414214 7.90336 0 7.90336C 0.414214 7.90336 0.75 7.56757 0.75 7.15336L -0.75 7.15336ZM 0.75 0C 0.75 -0.414214 0.414214 -0.75 0 -0.75C -0.414214 -0.75 -0.75 -0.414214 -0.75 0L 0.75 0ZM 0.75 7.15336L 0.75 0L -0.75 0L -0.75 7.15336L 0.75 7.15336Z"/>
<path id="path5_stroke" d="M 0 0L 0 -0.75C -0.414214 -0.75 -0.75 -0.414214 -0.75 1.66533e-16L 0 0ZM 3.26316 0L 4.01316 0C 4.01316 -0.414214 3.67737 -0.75 3.26316 -0.75L 3.26316 0ZM 3.26316 3.84814L 3.26316 4.59814C 3.67737 4.59814 4.01316 4.26235 4.01316 3.84814L 3.26316 3.84814ZM 0 3.84814L -0.75 3.84814C -0.75 4.26235 -0.414214 4.59814 1.66533e-16 4.59814L 0 3.84814ZM 0 0.75L 3.26316 0.75L 3.26316 -0.75L 0 -0.75L 0 0.75ZM 2.51316 0L 2.51316 3.84814L 4.01316 3.84814L 4.01316 0L 2.51316 0ZM 3.26316 3.09814L 0 3.09814L 0 4.59814L 3.26316 4.59814L 3.26316 3.09814ZM 0.75 3.84814L 0.75 0L -0.75 0L -0.75 3.84814L 0.75 3.84814Z"/>
</defs>
</svg>
