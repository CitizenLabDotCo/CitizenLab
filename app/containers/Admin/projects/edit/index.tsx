// Libraries
import * as React from 'react';
import * as Rx from 'rxjs/Rx';

// Services
import { projectBySlugStream, IProjectData } from 'services/projects';

// Components
import GoBackButton from 'components/UI/GoBackButton';
import TabbedResource from 'components/admin/TabbedResource';
import { browserHistory } from 'react-router';

// Localisation
import { InjectedIntlProps } from 'react-intl';
import { injectIntl } from 'utils/cl-intl';
import messages from '../messages';

// style
import styled from 'styled-components';

const StyledGoBackButton = styled(GoBackButton)`
  margin-top: 5px;
  margin-bottom: 30px;
`;

type Props = {
  params: {
    slug: string | null,
  },
  location: {
    pathname: string
  }
};

type State = {
  project: IProjectData | null,
  loaded: boolean
};

class AdminProjectEdition extends React.PureComponent<Props & InjectedIntlProps, State> {
  props$: Rx.BehaviorSubject<Props>;
  subscriptions: Rx.Subscription[];

  constructor(props: Props) {
    super(props as any);
    this.state = {
      project: null,
      loaded: false
    };
    this.props$ = new Rx.BehaviorSubject(null as any);
    this.subscriptions = [];
  }

  componentWillMount() {
    this.props$.next(this.props);

    this.subscriptions = [
      this.props$.distinctUntilChanged().switchMap((props) => {
        return (props.params && props.params.slug ? projectBySlugStream(props.params.slug).observable : Rx.Observable.of(null));
      }).subscribe((project) => {
        this.setState({
          project: (project ? project.data : null),
          loaded: true
        });
      })
    ];
  }

  componentWillReceiveProps(nextProps) {
    this.props$.next(nextProps);
  }

  componentWillUnmount() {
    this.subscriptions.forEach(subscription => subscription.unsubscribe());
  }

  getTabs = (slug: string, project: IProjectData) => {
    const baseTabsUrl = `/admin/projects/${slug}`;

    const tabs = [
      {
        label: this.props.intl.formatMessage(messages.generalTab),
        url: `${baseTabsUrl}/edit`,
      },
      {
        label: this.props.intl.formatMessage(messages.descriptionTab),
        url: `${baseTabsUrl}/description`,
      },
      {
        label: this.props.intl.formatMessage(messages.ideasTab),
        url: `${baseTabsUrl}/ideas`,
      },
      {
        label: this.props.intl.formatMessage(messages.eventsTab),
        url: `${baseTabsUrl}/events`,
      },
      {
        label: this.props.intl.formatMessage(messages.permissionsTab),
        url: `${baseTabsUrl}/permissions`,
        feature: 'private_projects',
      },
    ];

    if (project.attributes.process_type === 'timeline') {
      tabs.splice(3, 0, {
        label: this.props.intl.formatMessage(messages.phasesTab),
        url: `${baseTabsUrl}/timeline`,
      });
    }

    return tabs;
  }

  goBack = () => {
    const currentLocation = browserHistory.getCurrentLocation();
    const currentPath = currentLocation.pathname;
    const lastUrlSegment = currentPath.substr(currentPath.lastIndexOf('/') + 1);
    const newPath = currentPath.replace(lastUrlSegment, '').replace(/\/$/, '');
    const newLastUrlSegment = newPath.substr(newPath.lastIndexOf('/') + 1);

    if (newLastUrlSegment === this.props.params.slug) {
      browserHistory.push('/admin/projects');
    } else {
      browserHistory.push(newPath);
    }
  }

  render() {
    const { location } = this.props;
    const { slug } = this.props.params;
    const { project, loaded } = this.state;
    const { formatMessage } = this.props.intl;

    if (loaded) {
      const { children } = this.props;
      const childrenWithExtraProps = React.cloneElement(children as React.ReactElement<any>, { project });
      const tabbedProps = {
        location,
        resource: {
          title: project ? project.attributes.title_multiloc : formatMessage(messages.addNewProject),
          publicLink: project ? `/projects/${project.attributes.slug}` : ''
        },
        messages: {
          viewPublicResource: messages.viewPublicProject,
        },
        tabs: ((slug && project) ? this.getTabs(slug, project) : [])
      };

      return(
        <>
          <StyledGoBackButton onClick={this.goBack} />
          <TabbedResource {...tabbedProps}>
            {childrenWithExtraProps}
          </TabbedResource>
        </>
      );
    }

    return null;
  }
}

export default injectIntl<Props>(AdminProjectEdition);
