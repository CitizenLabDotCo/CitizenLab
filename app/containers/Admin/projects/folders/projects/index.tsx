import React, { Component } from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';
import { withRouter, WithRouterProps } from 'react-router';

// services
import { updateProjectFolderMembership, PublicationStatus } from 'services/projects';

// resources
import GetProjectFolder, { GetProjectFolderChildProps } from 'resources/GetProjectFolder';
import GetAdminPublications, { GetAdminPublicationsChildProps } from 'resources/GetAdminPublications';

// localisation
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

// components
import { List, Row, SortableList, SortableRow } from 'components/admin/ResourceList';
import { HeaderTitle } from '../../all/StyledComponents';
import ProjectRow from '../../components/ProjectRow';

// style
import styled from 'styled-components';
import { reorderAdminPublication } from 'services/adminPublications';
import { IAdminPublicationContent } from 'hooks/useAdminPublications';

const Container = styled.div`
  min-height: 60vh;
`;

const ListsContainer = styled.div``;

const ListHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  margin-bottom: 25px;
  &:not(:first-child) {
    margin-top: 70px;
  }
  & + & {
    margin-top: 30px;
  }
`;

const StyledHeaderTitle = styled(HeaderTitle)`
  font-weight: bold;
`;

const Spacer = styled.div`
  flex: 1;
`;

interface DataProps {
  topLevelProjects: GetAdminPublicationsChildProps;
  projectFolder: GetProjectFolderChildProps;
  projectsInFolder: GetAdminPublicationsChildProps;
}

interface Props extends DataProps { }

interface State {
  processing: string[];
}

class AdminFoldersProjectsList extends Component<Props & WithRouterProps, State> {
  constructor(props) {
    super(props);

    this.state = {
      processing: []
    };
  }

  handleReorder = (itemId, newOrder) => {
    reorderAdminPublication(itemId, newOrder);
  }

  addProjectToFolder = (projectId: string) => async () => {
    const projectFolderId = !isNilOrError(this.props.projectFolder) ? this.props.projectFolder.id : null;
    if (projectFolderId) {
      this.setState(({ processing }) => ({ processing: [...processing, projectId] }));
      await updateProjectFolderMembership(projectId, projectFolderId);
      this.setState(({ processing }) => ({ processing: processing.filter(item => item !== projectId) }));
    }
  }

  removeProjectFromFolder = (projectId: string) => async () => {
    const projectFolderId = !isNilOrError(this.props.projectFolder) ? this.props.projectFolder.id : undefined;
    this.setState(({ processing }) => ({ processing: [...processing, projectId] }));
    await updateProjectFolderMembership(projectId, null, projectFolderId);
    this.setState(({ processing }) => ({ processing: processing.filter(item => item !== projectId) }));
  }

  render() {
    const { topLevelProjects, projectsInFolder } = this.props;

    const { processing } = this.state;

    const otherProjects = (!isNilOrError(topLevelProjects) && topLevelProjects.list)
      ? topLevelProjects.list.filter(item => item.publicationType === 'project')
      : null;

    const inFolderFinalList = (!isNilOrError(projectsInFolder) && projectsInFolder.list && projectsInFolder.list.length > 0)
      ? projectsInFolder.list.filter(item => item.publicationType === 'project')
      : null;

    return (
      <Container>
        <ListsContainer>
          <ListHeader>
            <StyledHeaderTitle>
              <FormattedMessage {...messages.projectsAlreadyAdded} />
            </StyledHeaderTitle>

            <Spacer />

          </ListHeader>

          {inFolderFinalList ?
            <SortableList
              key={`IN_FOLDER_LIST${inFolderFinalList.length}`}
              items={inFolderFinalList}
              onReorder={this.handleReorder}
              className="projects-list e2e-admin-folder-projects-list"
              id="e2e-admin-fodlers-projects-list"
            >
              {({ itemsList, handleDragRow, handleDropRow }) => (
                itemsList.map((adminPublication: IAdminPublicationContent, index) => {
                  return (
                    <SortableRow
                      key={adminPublication.id}
                      id={adminPublication.id}
                      index={index}
                      moveRow={handleDragRow}
                      dropRow={handleDropRow}
                      lastItem={(index === itemsList.length - 1)}
                    >
                      <ProjectRow
                        publication={adminPublication}
                        actions={[{
                          buttonContent: <FormattedMessage {...messages.removeFromFolder} />,
                          handler: this.removeProjectFromFolder,
                          icon: 'remove',
                          processing: processing.includes(adminPublication.publicationId)
                        }, 'manage']}
                      />
                    </SortableRow>
                  );
                })
              )}
            </SortableList>
            :
            <FormattedMessage {...messages.emptyFolder} />
          }
          <ListHeader>
            <StyledHeaderTitle>
              <FormattedMessage {...messages.projectsYouCanAdd} />
            </StyledHeaderTitle>
          </ListHeader>

          {otherProjects ?
            <List key={`JUST_LIST${otherProjects.length}`}>
              {otherProjects.map((adminPublication, index: number) => {
                return (
                  <Row
                    id={adminPublication.id}
                    lastItem={(index === otherProjects.length - 1)}
                    key={adminPublication.id}
                  >
                    <ProjectRow
                      publication={adminPublication}
                      actions={[{
                        buttonContent: <FormattedMessage {...messages.addToFolder} />,
                        handler: this.addProjectToFolder,
                        processing: processing.includes(adminPublication.publicationId),
                        icon: 'plus-circle'
                      }]}
                    />
                  </Row>
                );
              })}
            </List>
            : <FormattedMessage {...messages.noProjectsToAdd} />
          }
        </ListsContainer>
      </Container>
    );
  }
}
const AdminFoldersProjectsListWithHocs = withRouter(AdminFoldersProjectsList);

const publicationStatuses: PublicationStatus[] = ['draft', 'archived', 'published'];

const Data = adopt<DataProps, WithRouterProps>({
  projectFolder: ({ params, render }) => <GetProjectFolder projectFolderId={params.projectFolderId}>{render}</GetProjectFolder>,
  topLevelProjects: <GetAdminPublications publicationStatusFilter={publicationStatuses} folderId={null} />,
  projectsInFolder: ({ params, render }) => <GetAdminPublications publicationStatusFilter={publicationStatuses} folderId={params.projectFolderId}>{render}</GetAdminPublications>,
});

export default (inputProps: WithRouterProps) => (
  <Data {...inputProps}>
    {dataProps => <AdminFoldersProjectsListWithHocs {...inputProps} {...dataProps} />}
  </Data>
);
