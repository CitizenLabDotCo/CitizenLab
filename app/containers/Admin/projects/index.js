import React from 'react';
import PropTypes from 'prop-types';
import HelmetIntl from 'components/HelmetIntl';

// components
// import { FormattedMessage } from 'react-intl';
import WatchSagas from 'utils/containers/watchSagas';
import sagas from 'resources/projects/sagas';

// store
import { preprocess } from 'utils';

// messages
import messages from './messages';
import { resetProjects } from 'resources/projects/actions';

class ProjectDashboard extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function

  componentWillUnmount() {
    // reset projects upon page leaving, to avoid double items when going back
    this.props.resetProjects();
  }

  render() {
    return (
      <div>
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        <WatchSagas sagas={sagas} />
        {this.props.children}
      </div>

    );
  }
}

ProjectDashboard.propTypes = {
  resetProjects: PropTypes.func.isRequired,
  children: PropTypes.element,
};

export default preprocess(null, { resetProjects })(ProjectDashboard);
