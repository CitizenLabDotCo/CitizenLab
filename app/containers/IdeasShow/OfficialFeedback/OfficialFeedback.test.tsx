import React from 'react';

import { shallow } from 'enzyme';

import { getMockProject } from 'services/projects';
import { mockOfficialFeedback } from 'services/officialFeedback';

// import OfficialFeedbackNew from './Form/OfficialFeedbackNew';

jest.mock('./Form/OfficialFeedbackNew', () => 'hi');
jest.mock('services/projects');
jest.mock('services/officialFeedback');
jest.mock('resources/GetPermission');

import 'jest-styled-components';

const mockProject = getMockProject('projectId', 'continuous', 'ideation');

const mockOfficialFeedbackChild = { officialFeedbackList: mockOfficialFeedback.data };

import { OfficialFeedback } from './';

describe('<OfficialFeedback />', () => {
  it('renders correctly for non-admin', () => {
    const wrapper = shallow(<OfficialFeedback ideaId="ideaId" project={mockProject} permission={false} officialFeedback={mockOfficialFeedbackChild} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly for admin', () => {
    const wrapper = shallow(<OfficialFeedback ideaId="ideaId" project={mockProject} permission={true} officialFeedback={mockOfficialFeedbackChild} />);
    expect(wrapper).toMatchSnapshot();
  });
});
