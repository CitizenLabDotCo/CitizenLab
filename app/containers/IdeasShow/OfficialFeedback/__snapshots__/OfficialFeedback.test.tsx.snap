// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfficialFeedback /> renders correctly for admin 1`] = `
<Fragment>
  <OfficialFeedback__StyledOfficialFeedbackNew
    ideaId="ideaId"
  />
  <OfficialFeedback__FeedbackHeader>
    <OfficialFeedback__FeedbackTitle>
      <FormattedMessage
        defaultMessage="Official updates"
        id="app.components.IdeasShow.OfficialFeedback.officialUpdates"
      />
    </OfficialFeedback__FeedbackTitle>
    <FormattedMessage
      defaultMessage="Last update: {lastUpdateDate}"
      id="app.components.IdeasShow.OfficialFeedback.lastUpdate"
      values={
        Object {
          "lastUpdateDate": <OfficialFeedback__StyledSpan>
            <FormattedDate
              value="2011-01-01T04:26:07.000Z"
            />
          </OfficialFeedback__StyledSpan>,
        }
      }
    />
  </OfficialFeedback__FeedbackHeader>
  <WrappedOfficialFeedBack
    editingAllowed={true}
    ideaId="ideaId"
  />
</Fragment>
`;

exports[`<OfficialFeedback /> renders correctly for non-admin 1`] = `
<Fragment>
  <OfficialFeedback__FeedbackHeader>
    <OfficialFeedback__FeedbackTitle>
      <FormattedMessage
        defaultMessage="Official updates"
        id="app.components.IdeasShow.OfficialFeedback.officialUpdates"
      />
    </OfficialFeedback__FeedbackTitle>
    <FormattedMessage
      defaultMessage="Last update: {lastUpdateDate}"
      id="app.components.IdeasShow.OfficialFeedback.lastUpdate"
      values={
        Object {
          "lastUpdateDate": <OfficialFeedback__StyledSpan>
            <FormattedDate
              value="2011-01-01T04:26:07.000Z"
            />
          </OfficialFeedback__StyledSpan>,
        }
      }
    />
  </OfficialFeedback__FeedbackHeader>
  <WrappedOfficialFeedBack
    editingAllowed={false}
    ideaId="ideaId"
  />
</Fragment>
`;
